guide/src/class/numeric.md

```diff
@@ -0,0 +1,373 @@
+
+### Fixing our constructor
+
+Let's address the first overflow, in `Number`'s constructor:
+
+```python
+from my_module import Number
+
+n = Number(1 << 1337)
+```
+
+```text
+Traceback (most recent call last):
+  File "example.py", line 3, in <module>
+    n = Number(1 << 1337)
+OverflowError: Python int too large to convert to C long
+```
+
+Instead of relying on the default [`FromPyObject`] extraction to parse arguments, we can specify our
+own extraction function, using the `#[pyo3(from_py_with = "...")]` attribute. Unfortunately PyO3
+ doesn't provide a way to wrap Python integers out of the box, but there's a Python C API function that does:
+
+```c
+unsigned long PyLong_AsUnsignedLongMask(PyObject *obj)
+```
+
+We can call this function from Rust by using [`pyo3::ffi::PyLong_AsUnsignedLongMask`]. This is an *unsafe*
+function, which means we have to use an unsafe block to call it and take responsibility for upholding
+the contracts of this function. Let's review those contracts:
+- The GIL must be held. If it's not, calling this function causes a data race.
+- The pointer must be valid, i.e. it must be properly aligned and point to a valid Python object.
+
+Let's create that helper function. The signature has to be `fn(&PyAny) -> PyResult<T>`.
+- `&PyAny` represents a checked borrowed reference, so the pointer derived from it is valid (and not null).
+- Whenever we have borrowed references to Python objects in scope, it is guaranteed that the GIL is held. This reference is also where we can get a [`Python`] token to use in our call to [`PyErr::take`].
+
+We also add documentation, via `///` comments and the `#[pyo3(text_signature = "...")]` attribute, both of which are visible to Python users.
+
+```rust
+use std::os::raw::c_ulong;
+
+use pyo3::prelude::*;
+use pyo3::conversion::AsPyPointer;
+use pyo3::ffi;
+
+fn wrap(obj: &PyAny) -> Result<i32, PyErr> {
+    let py: Python = obj.py();
+
+    unsafe {
+        let ptr = obj.as_ptr();
+
+        let ret: c_ulong = ffi::PyLong_AsUnsignedLongMask(ptr);
+
+        //if it returns max we have to check whether an error occurred.
+        if ret == c_ulong::MAX {
+            if let Some(err) = PyErr::take(py) {
+                return Err(err);
+            }
+        }
+
+        Ok(ret as i32)
+    }
+}
+
+/// Did you ever hear the tragedy of Darth Signed The Overfloweth? I thought not.
+/// It's not a story C would tell you. It's a Rust legend.
+#[pyclass(module = "my_module")]
+#[pyo3(text_signature = "(int)")]
+struct Number(i32);
+
+#[pymethods]
+impl Number {
+    #[new]
+    fn new(#[pyo3(from_py_with = "wrap")] value: i32) -> Self {
+        Self(value)
+    }
+}
+```
+
+
+With that out of the way, let's implement some operators:
+```rust
+use std::convert::TryInto;
+use pyo3::exceptions::{PyZeroDivisionError, PyValueError};
+
+# use pyo3::prelude::*;
+#
+# #[pyclass]
+# struct Number(i32);
+#
+#[pymethods]
+impl Number {
+    fn __add__(&self, other: &Self) -> Self {
```

Comment:
Aside: because this type wraps `i32` it's probably slightly more efficient to pass by `self`. But I don't think we support that. Might be nice to do so (maybe for `Copy` types only)?

Reply:
I think even if that was supported we shouldn't do it, because if users copy code and try that with non-copy types it'll throw some cryptic compiler errors.

Reply:
Ok 👍